pkg/proctree/proctree.go

```diff
@@ -0,0 +1,325 @@
+
+	if childPid == childTid {
+		// The child is a process, not a thread.
+		child := pt.GetOrCreateProcessByHash(childHash)
+
+		childInfo := child.GetInfo()
+		childInfo.SetTid(int(childTid)) // TODO: optmization: change all fields at once
+		childInfo.SetPid(int(childPid))
+		childInfo.SetNsTid(int(childNsTid))
+		childInfo.SetNsPid(int(childNsPid))
+		childInfo.SetStartTime(childStartTime)
+		childInfo.SetPPid(int(parentPid))
+		childInfo.SetNsPPid(int(parentNsPid))
+		childInfo.SetUid(0) // TODO: implement
+		childInfo.SetGid(0) // TODO: implement
+
+		parentInfo := parent.GetInfo()
+		parentInfo.SetTid(int(parentTid))
+		parentInfo.SetPid(int(parentPid))
+		parentInfo.SetNsTid(int(parentNsTid))
+		parentInfo.SetNsPid(int(parentNsPid))
+		parentInfo.SetStartTime(parentStartTime)
+
+		child.SetParentHash(parentHash)
+		parent.AddChild(childHash)
+	} else {
+		// The child is a thread, not a process.
+
+		// The importance of keeping records for threads is this: Whenever there is an event, in the
+		// regular event pipeline, related to this thread, tracee can find the parent of the thread
+		// (the process responsible for spawning the thread group) and tie the event to the real
+		// process (files or sockets opened, resources accessed, and so on).
+
+		child := pt.GetOrCreateThreadByHash(childHash)
+		child.SetParentHash(parentHash)
+		parent.addThread(childHash)
+	}
+
+	// DEBUG (TODO: remove this)
+	// file, _ := os.Open("/dev/null")
+	// file := os.Stdout
+	// fmt.Fprintf(file, "--\nFork event received:\n")
+	// fmt.Fprintf(file, "Parent Hash: %v\n", parentHash)
+	// fmt.Fprintf(file, "Parent: tid=%v ns_tid=%v pid=%v ns_pid=%v start_time=%v\n", parentTid, parentNsTid, parentPid, parentNsPid, parentStartTime)
+	// fmt.Fprintf(file, "Task Hash: %v\n", childHash)
+	// fmt.Fprintf(file, "Child: tid=%v ns_tid=%v pid=%v ns_pid=%v start_time=%v\n", childTid, childNsTid, childPid, childNsPid, childStartTime)
+	// END OF DEBUG
+
+	return nil
+}
+
+// FeedFromExec feeds the process tree with an exec event.
+func (pt *ProcessTree) FeedFromExec(
+	taskHash uint32,
+	cmdPath string,
+	pathName string,
+	dev uint32,
+	inode uint64,
+	ctime uint64,
+	inodeMode uint16,
+	interPathName string,
+	interDev uint32,
+	interInode uint64,
+	interCtime uint64,
+	inter string,
+	stdinType uint16,
+	stdinPath string,
+	invokedFromKernel int32,
+) error {
+	process, procOk := pt.GetProcessByHash(taskHash)
+	thread, threadOk := pt.GetThreadByHash(taskHash)
+
+	if !procOk && !threadOk {
+		logger.Debugw("process or thread not found (evicted ?)", "taskHash", taskHash)
+		return nil
+	}
+
+	// Running execve() from a thread is discouraged and behavior can be unexpected:
+	//
+	// 1. All threads are terminated.
+	// 2. PID remains the same.
+	// 3. New process is single threaded.
+	// 4. Inherited Attributed (open fds, proc group, ID, UID, GID, ... are retained)
+	// 5. Still, it isn't forbidden and should be handled.
```

Comment:
You mean to zero the "threads []" slice in the process (and keep the history ?). Yep, I think thats right. Because, technically, it is the same process (so the hash will be the same) but we need history about what happened.

Reply:
Yea exactly.
However, the main thread will live after the `execve`, so the threads slice should contain it (instead of being empty).

Reply:
Still a TODO as this is a corner case. I believe the Exit() will be called (and a controller/signal event sent) to each thread that has exited. I need to create a test case for this particular case, will do after this PR.

Reply:
https://github.com/aquasecurity/tracee/issues/3478